src/dune/lib.ml

```diff
@@ -965,11 +982,57 @@ end = struct
       loop Map.empty (Map.to_list impls)
   end
 
+  (* For a proper linking closure, we must make sure that an implementation
+     cannot depend on a library that has the virtual library as a dependency *)
+  let validate_closure ~vlib ~impl impls =
+    let rec lib (l : lib) ~stack =
+      if l = vlib then
+        Error.make
+          [ Pp.textf
+              "Implementation %s has a dependency that uses the virtual \
+               library %s"
+              (Lib_name.to_string impl.name)
+              (Lib_name.to_string vlib.name)
+          ; Dep_stack.pp stack
+          ]
+      else
+        match Lib_info.virtual_ l.info with
+        | None ->
+          let* stack = Dep_stack.push stack ~implements_via:None l.unique_id in
+          requires l.requires ~stack
+        | Some _ -> (
+          let* () =
+            let* stack =
+              Dep_stack.push stack ~implements_via:None l.unique_id
+            in
+            requires l.requires ~stack
+          in
+          match Map.find impls l with
+          | None -> Ok ()
+          | Some l ->
+            let* stack =
+              Dep_stack.push stack ~implements_via:None l.unique_id
+            in
+            requires l.requires_no_vlib ~stack )
+    and requires deps ~stack = deps >>= Result.List.iter ~f:(lib ~stack) in
+    let* stack =
+      Dep_stack.push Dep_stack.empty ~implements_via:None impl.unique_id
+    in
+    requires impl.requires_no_vlib ~stack
+
   let second_step_closure ts impls =
     let visited = ref Id.Set.empty in
+    let find_impl (lib : lib) =
+      match Lib_info.virtual_ lib.info with
+      | None -> Ok lib
+      | Some _ ->
+        let impl = Map.find_exn impls lib in
```

Comment:
Can this raise? If yes, we should go through the `Or_exn.t` monad. If not, we should had a comment justifying why this cannot raise.

Reply:
It cannot raise. I added an appropriate comment.